network/src/traits.rs

```diff
@@ -104,3 +105,80 @@ pub trait ReliableNetwork: BaseNetwork {
         handlers
     }
 }
+
+#[async_trait]
+pub trait UnreliableNetwork2<Message: Clone + Send + Sync> {
+    async fn unreliable_send(&mut self, peer: PublicKey, message: &Message) -> JoinHandle<()>;
+
+    /// Broadcasts a message to all `peers` passed as an argument.
+    /// The attempts to send individual messages are best effort and will not be retried.
+    async fn unreliable_broadcast(
+        &mut self,
+        peers: Vec<PublicKey>,
+        message: &Message,
+    ) -> Vec<JoinHandle<()>> {
+        let mut handlers = Vec::new();
+        for peer in peers {
+            let handle = { self.unreliable_send(peer, message).await };
+            handlers.push(handle);
+        }
+        handlers
+    }
+}
+
+pub trait Lucky {
+    fn rng(&mut self) -> &mut SmallRng;
+}
+
+#[async_trait]
+pub trait LuckyNetwork2<Message> {
+    /// Pick a few addresses at random (specified by `nodes`) and try (best-effort) to send the
+    /// message only to them. This is useful to pick nodes with whom to sync.
+    async fn lucky_broadcast(
+        &mut self,
+        mut peers: Vec<PublicKey>,
+        message: &Message,
+        nodes: usize,
+    ) -> Vec<JoinHandle<()>>;
+}
+
+#[async_trait]
+impl<T, M> LuckyNetwork2<M> for T
+where
+    M: Clone + Send + Sync,
+    T: UnreliableNetwork2<M> + Send,
+    T: Lucky,
+{
+    async fn lucky_broadcast(
+        &mut self,
+        mut peers: Vec<PublicKey>,
+        message: &M,
+        nodes: usize,
+    ) -> Vec<JoinHandle<()>> {
+        peers.shuffle(self.rng());
```

Comment:
Would `partial_shuffle` be better here to avoid shuffling stuff you're just gonna truncate?

(Or `choose_multiple`, which seems more natural, but would require reworking all these interfaces to use iterators instead of vecs? Curious if that's worth considering, or if not then how did you come to decide to pass around vecs?)

Reply:
This code was really just moved I don't know the exact motivation for how it was done here. I think this works for now but we can revisit and clean it up once the networking migration has finished.